src/common/AdaptiveStream.cpp

```diff
@@ -720,15 +722,32 @@ bool AdaptiveStream::start_stream(uint64_t startPts)
             current_rep_->get_segment(static_cast<size_t>(segmentId - current_rep_->GetStartNumber()));
       }
     }
+    else if (stream_changed_) // switching streams, align new stream segment no.
+    {
+      uint64_t segmentId = segment_buffers_[0]->segment_number;
+      if (segmentId >= current_rep_->GetStartNumber() + current_rep_->SegmentTimeline().GetSize())
+      {
+        segmentId = current_rep_->GetStartNumber() + current_rep_->SegmentTimeline().GetSize() - 1;
+      }
+      current_rep_->current_segment_ = current_rep_->get_segment(
+          static_cast<size_t>(segmentId - current_rep_->GetStartNumber()));
+    }
     else
+    {
       current_rep_->current_segment_ = nullptr; // start from beginning
+    }
   }
 
   const CSegment* next_segment =
       current_rep_->get_next_segment(current_rep_->current_segment_);
 
   if (!next_segment)
   {
+    //! @todo: THIS MUST BE CHANGED - !! BUG !!
+    //! this will broken/stop playback on live streams when adaptive stream change stream quality (so representation)
+    //! and there are no new segments because not immediately available (despite child manifest updated),
+    //! because we cant be ensure next segment without wait the appropriate timing...imo its not the right place here.
+    //! Can be reproduced with HLS live and by using "Stream selection" setting to "Test" by switching per 1 segment
```

Comment:
yes the scenario makes sense why it fails. We could try a hacky workaround until something better comes along - assume when the next segment comes it will be the same size (usually is) and add that PTS to current segment's start PTS. Sound good or not?

Reply:
i understood but as you said its hacky im not 100% sure that could be so good

i found this problem even when trying to fix bugs to make this live sample playable
https://livesim.dashif.org/livesim/periods_20/testpic_2s/Manifest.mpd

This live sample have periods but no new segments provided,
then its required the use of "InsertLiveSegment" in conjuction with the manifest update worker.
This sample somewhat works on kodi <=v20 because of old code hacks and periods seem "ignored",
on kodi 21 is ofc broken, from what i see seem there are multiple problems on  `CDashTree::RefreshLiveSegments` that seem dont take in account of "inserted" segments on timeline, delete them all and so seem to loose the last inserted segment with the updated pts, this will make problems with `AdaptiveStream::ensureSegment` that its not able to find next segment, `AdaptiveStream::start_stream` have problem when switch period, and also with `CSession::OnSegmentChangedRead` that its not able to insert new segments anymore
(these problems may be also similar with https://github.com/xbmc/inputstream.adaptive/issues/1461)